### ISSUE TITLE
Covariant Lyapunov Vectors from data using SVD 

### ISSUE THREAD

**Issue by Datseris** [JuliaDynamics/ChaosTools.jl#308]
Covariant Lyapunov vectors are related with Lyapunov exponents and in essense show the direction of increase/decrease each exponent quantifies. This paper

![image](https://github.com/JuliaDynamics/ChaosTools.jl/assets/19669089/98e9b66f-9123-4017-81bd-de63f3d3dcee)

https://www.sciencedirect.com/science/article/pii/S0167278912003090

provides some rather simple algorithms to estimate CLVs from a data matrix (paper also exists on arxiv). The paper also has MatLab code for the computation. @CourtneyQuinn was kind enough to provide Python implementation as well here: https://github.com/CourtneyQuinn/FEM-BV-VAR_dynamics/blob/v0.1.0/src/clustering_dynamics/dynamics/clvs.py algorithm 2.2 is what we are interested in.

It is honestly amazing how simple the implementation is :D 

But I wonder, are there any easy to implement algorithms to implement the CLVs from a dynamical systems, not from data? So using our `TangentDynamicalSystem`?

**Comment by hinsley** [comment]
> But I wonder, are there any easy to implement algorithms to implement the CLVs from a dynamical systems, not from data? So using our `TangentDynamicalSystem`?

Yes, there is the [Ginelli algorithm](https://arxiv.org/abs/0706.0510) and the [Wolfe & Samelson](https://chaosbk.physics.gatech.edu/library/WoSa07.pdf) algorithm, both published in 2007. An article by [Kuptsov and Parlitz](https://arxiv.org/abs/1105.5228) summarizes both of these and is probably the best resource, also providing its own algorithm.

**Ginelli's algorithm:**
- Very similar to the QR algorithm for Lyapunov exponents; you save the $R$ matrices at each point on a trajectory and the vectors are then evolved backward from the end of the trajectory with the inverse tangent system (still normalizing every few steps, but not orthogonalizing); they rather quickly converge to the CLVs. You can even compute Lyapunov exponents during the CLV computation.

**Wolfe & Samelson algorithm:**
- I haven't tried to implement this myself yet, but it seems like you don't need to save a bunch of upper-triangular $R$ matrices from an initial forward pass, so it's more memory efficient.
- It seems like this algorithm is based on singular vectors / some kind of subspace intersection properties and has some guarantees about dimensionality (for instance, for computing the leading $k$ singular vectors at a certain point you only need $2k-1$ leading singular vectors, $k-1$ being forward vectors and $k$ being backward vectors). For high-dimensional systems this guarantee is very helpful. Ginelli's method may allow efficient computation of only a subset of CLVs too, but I haven't put any thought into it.

The nice thing about CLVs is that they provide a way to quickly check for dominated splittings for e.g. determining (pseudo)hyperbolicity or existence of homoclinic tangencies in maps and flows, though these algorithms are only asymptotically convergent so they can't rigorously justify such properties. There is also a possibility of using CLVs in bifurcation analysis, detecting period-doubling bifurcations of limit cycles and inclination flip bifurcations of homoclinic trajectories, though such approaches are likely more expensive than standard approaches in MATCONT and BifurcationKit.jl.

Since I've already implemented the Ginelli algorithm in some toy research projects, I'd be happy to try contributing an implementation to ChaosTools.jl for use with `TangentDynamicalSystem`. I should note that there is another existing Julia implementation of the Ginelli algorithm [here](https://github.com/tkf/LyapunovExponents.jl).

EDIT: I submitted PR #357 implementing this feature.
